### PR TITLE
refactor: load and read logic

### PR DESCRIPTION
**Pull request by teodora-sandu**
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/CONTRIBUTING.md) rules

#### What does this PR do?

Refactors the load/parse files functionality in the IaC test flow so it's easier to follow and both non-Terraform and Terraform files (post `iacTerraformVarSupport` feature flag) use the same directory and file reading functionality.

#### Where should the reviewer start?
Start in `index.ts` and follow the flow through. I've added comments throughout the PR but if it's difficult to read then please ask for a Zoom.

Hammer and Open Source accidentally own a file that wasn't added to our owernship: `test/jest/unit/makeDirectoryIterator.spec.ts`

#### Any background context you want to provide?
We realised that the current flow is very hard to follow and expand. The non-Terraform files would use one file-reading functionality, parse its files, then we'd move onto Terraform files to use another file-reading functionality. Now, we want to use the same file-reading functionality (the one introduced by the variable support work) for both the Terraform and non-Terraform files.

Current flow:
- separate TF from non-TF files if TF Vars support is enabled
- for non-TF files:
    - `loadFiles`
         - `loadContentForFiles`
    - `parseFiles`(parses each file individually)
- for TF files:
    - `getAllDirectoriesForPath`:
    -  for each directory:
         - `getFilesForDirectory`
         - `loadContentForFiles`
         - `parseTfFiles`(parses multiple files at once)

New flow:
 - `getAllDirectoriesForPath`:
    -  for each directory
         - `getFilesForDirectory`
         - `loadContentForFiles`
         - `parseFiles`:
             - separate TF from non-TF files if TF Vars support is enabled
             - for non-TF files:
                 - `parseFiles`(parses each file individually)
             - for TF files:
                   - `parseTfFiles`(parses multiple files at once)

### CODE REVIEW

File: src/cli/commands/test/iac-local-execution/file-loader.ts [hunk 1]
```diff
@@ -28,60 +21,6 @@ export async function loadContentForFiles(
...
-    maxDepth: options.detectionDepth,
-    validFileTypes,
-  });
-
-  if (filePaths.length === 0) {
-    throw new NoFilesToScanError();
-  }
-  const loadedFiles = await loadContentForFiles(filePaths);
-
-  return loadedFiles.filter((file) => file.fileContent !== '');
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/file-loader.ts:46]: TODO

File: src/cli/commands/test/iac-local-execution/file-loader.ts [hunk 2]
```diff
@@ -28,60 +21,6 @@ export async function loadContentForFiles(
...
 }
 
-export async function loadFiles(
-  pathToScan: string,
-  options: IaCTestFlags = {},
-  validFileTypes?: string[],
-): Promise<IacFileData[]> {
-  const filePaths = getFilePathsFromDirectory(pathToScan, {
-    maxDepth: options.detectionDepth,
-    validFileTypes,
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/file-loader.ts:38]: TODO

File: src/lib/iac/makeDirectoryIterator.ts [hunk 3]
```diff
@@ -1,54 +0,0 @@
...
-  }
-
-  // Internal function to allow us to track the recursion depth.
-  function* walk(dirpath: string, currentDepth: number) {
-    const filenames = fs.readdirSync(dirpath);
-    for (const filename of filenames) {
-      // NOTE: We filter filenames that start with a period to maintain
-      // backwards compatibility with the original implementation that used the
-      // "glob" package.
-      if (!options.includeDotfiles && filename.startsWith('.')) {
```
Comment (teodora-sandu) [inline src/lib/iac/makeDirectoryIterator.ts:24]: TODO

File: src/cli/commands/test/iac-local-execution/file-parser.ts [hunk 4]
```diff
@@ -32,24 +29,56 @@ const debug = Debug('snyk-test');
...
       parsedFiles.push(...tryParseIacFile(fileData, options));
     } catch (err) {
-      if (filesData.length === 1) {
-        throw err;
-      }
       failedFiles.push(generateFailedParsedFile(fileData, err));
     }
   }
 
-  if (parsedFiles.length === 0) {
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/file-parser.ts:49]: Moved to `index.ts` so we can deal with the case where TF and non-TF files are parsed separately

File: src/cli/commands/test/iac-local-execution/directory-loader.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+ * @param {currentDirectory} currentDirectory - The directory for the filepath
+ * @returns {boolean} if the filepath should be ignored or not
+ */
+export function isDotFile(
+  pathToScan: string,
+  currentDirectory: string,
+): boolean {
+  // we resolve the path in case the user tries to scan a single file with a relative path
+  // e.g. './my-folder/terraform.tf', or '../my-folder/terraform.tf'
+  const resolvedPath = path.resolve(currentDirectory, currentDirectory);
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/directory-loader.ts:143]: `path.resolve` instead of `path.normalize` to catch the `../my-folder/terraform.tf` edge case which was causing the GitHub action tests to fail

File: src/cli/commands/test/iac-local-execution/file-loader.ts [hunk 6]
```diff
@@ -28,60 +21,6 @@ export async function loadContentForFiles(
...
-
-  if (filePaths.length === 0) {
-    throw new NoFilesToScanError();
-  }
-  const loadedFiles = await loadContentForFiles(filePaths);
-
-  return loadedFiles.filter((file) => file.fileContent !== '');
-}
-
-function hasValidFileType(
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/file-loader.ts:49]: Now `shouldBeParsed` in `src/cli/commands/test/iac-local-execution/directory-loader.ts`

File: src/cli/commands/test/iac-local-execution/handle-terraform-files-utils.ts [hunk 7]
```diff
@@ -1,68 +0,0 @@
...
-            },
-          };
-        }
-      }
-    }
-  }
-  yield* generatorHelper(root, 1);
-}
-
-export const getExtensionForPath = (pathToScan: string) =>
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/handle-terraform-files-utils.ts:43]: Replaced with existing `getFileType`

File: src/cli/commands/test/iac-local-execution/handle-terraform-files-utils.ts [hunk 8]
```diff
@@ -1,68 +0,0 @@
...
-  path.extname(pathToScan).toLowerCase();
-
-export const shouldBeParsed = (pathToScan: string): boolean =>
-  VALID_TERRAFORM_FILE_TYPES.includes(getExtensionForPath(pathToScan));
-
-export const isSingleFile = (pathToScan: string): boolean =>
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/handle-terraform-files-utils.ts:49]: Replaced with existing `!isLocalFolder`

File: src/cli/commands/test/iac-local-execution/directory-loader.ts [hunk 9]
```diff
@@ -1,55 +1,6 @@
...
-
-/**
- * This function handles everything from filtering out Terraform files  directory by directory,
- * loading them, sending them to the parser and getting the parsing results back.
- * Then it concatenates and returns the new parsing results with the existing parsing results.
- * @param pathToScan - the path to scan provided by the user
- * @param maxDepth? - an optional maxDepth of directories if provided by the detection-depth flag
- * @returns { parsedFiles, failedFiles} - all the parsing results so far
- */
-export async function loadAndParseTerraformFiles(
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/directory-loader.ts:20]: Moved to `file-loader.ts` for better separation

File: src/cli/commands/test/iac-local-execution/index.ts [hunk 10]
```diff
@@ -57,73 +58,58 @@ export async function test(
...
-        !validFileTypes.includes(ValidFileType.Terraform) &&
-        err instanceof NoFilesToScanError
-      ) {
-        // ignore this error since we might only have .tf files in the folder and we will parse them in the next block
-      } else {
-        throw err;
-      }
+    // we load and parse files directory by directory
+    // because we need all files in the same directory to share the same variable context for Terraform
+    for (const currentDirectory of allDirectories) {
```
Comment (teodora-sandu) [inline src/cli/commands/test/iac-local-execution/index.ts:77]: I will move this into another file in a future PR but I didn't want to make this PR even harder to review...

File: test/jest/unit/iac/file-parser.spec.ts [hunk 11]
```diff
@@ -95,31 +90,38 @@ describe('parseFiles', () => {
...
-    await expect(
-      parseFiles([
-        {
-          fileContent: 'file.java',
-          filePath: 'path/to/file',
-          fileType: 'java' as IacFileTypes,
-        },
-      ]),
-    ).rejects.toThrow(UnsupportedFileTypeError);
+  it('includes unsupported file types in failed files', async () => {
```
Comment (teodora-sandu) [inline test/jest/unit/iac/file-parser.spec.ts:93]: These cases are also covered by the acceptance tests so this just verified the behaviour of `parseFiles`, which now has changed

File: test/jest/unit/iac/file-utils.spec.ts [hunk 12]
```diff
@@ -36,3 +40,48 @@ describe('extractBundle', () => {
...
+  it('limits the subdirectory depth when maxDepth option is provided', () => {
+    const it = makeFileAndDirectoryGenerator(fixturePath, 3);
+    const files = Array.from(it)
+      .filter((result) => result.file?.fileName !== undefined)
+      .map((result) => result.file?.fileName);
+
+    expect(files).toEqual([
+      expect.stringContaining('hidden.tf'),
+      expect.stringContaining('.hidden.tf'),
+      expect.stringContaining('one.tf'),
```
Comment (teodora-sandu) [inline test/jest/unit/iac/file-utils.spec.ts:78]: This stops one folder higher than the older function. We either include one more level and remain backwards compatible with the old or we stop earlier and remain backwards compatible with the new Terraform variable support work. I think we should stick to the old code but what do you think @ipapast ? 

Based on the docs I would say probably the old code was correct: 
Screenshot 2022-04-04 at 18 16 55

Comment (ipapast) [reply test/jest/unit/iac/file-utils.spec.ts:78]: ha, I think this is my mistake not knowing (and checking it) that detection depth is (current + depth). This was probably a product decision in the past and we should stick to what it was.

File: src/cli/commands/test/iac-local-execution/directory-loader.ts [hunk 13]
```diff
@@ -119,9 +73,35 @@ export function* getTerraformFilesInDirectoryGenerator(
...
   }
 }
+
+export const shouldBeParsed = (pathToScan: string): boolean =>
+  VALID_FILE_TYPES.includes(getFileType(pathToScan));
+
+export function getFileType(filePath: string): string {
+  const filePathSplit = filePath.split('.');
+  return filePathSplit[filePathSplit.length - 1].toLowerCase();
+}
```
Comment (ipapast) [inline src/cli/commands/test/iac-local-execution/directory-loader.ts:89]: This function is error prune if you pass different paths such as no dot, or other cases. I had created a new function in the handle-terraform-utils that would handle this slightly better with `path`:

https://github.com/snyk/cli/blob/bf8281ad450f6eb4658d4f44c8167b94de5d4677/src/cli/commands/test/iac-local-execution/handle-terraform-files-utils.ts#L43-L44

the other option is to manually guard it like this: https://github.com/snyk/cli/blob/336828deda1728ea42422db8e061df782453be2c/src/lib/iac/iac-parser.ts#L22-L30

File: src/cli/commands/test/iac-local-execution/file-parser.ts [hunk 14]
```diff
@@ -32,24 +29,56 @@ const debug = Debug('snyk-test');
 export async function parseFiles(
   filesData: IacFileData[],
   options: IaCTestFlags = {},
+  isTFVarSupportEnabled = false,
 ): Promise<ParsingResults> {
+  let tfFilesData: IacFileData[] = [];
+  let nonTfFilesData: IacFileData[] = [];
```
Comment (ipapast) [inline src/cli/commands/test/iac-local-execution/file-parser.ts:35]: I can refactor it later, but it will be easier if it's done by a bulk rename: 😬 
```suggestion
  let tfFileData: IacFileData[] = [];
  let nonTfFileData: IacFileData[] = [];
```
so that the name follows the type: singular describing a noun

File: src/cli/commands/test/iac-local-execution/directory-loader.ts (hunk 13 above)
Comment (teodora-sandu) [reply src/cli/commands/test/iac-local-execution/directory-loader.ts:89]: Thanks, I wasn't aware of that! I've ended up using your implementation but with a small change so we don't need to include the `.` in the beginning

File: src/cli/commands/test/iac-local-execution/file-utils.ts [hunk 15]
```diff
@@ -101,3 +103,42 @@ export function computePaths(
...
+          };
+        }
+      }
+    }
+  }
+  yield* generatorHelper(root, 0);
+}
+
+export const isSingleFile = (pathToScan: string): boolean =>
+  !isLocalFolder(pathToScan);
```
Comment (ipapast) [inline src/cli/commands/test/iac-local-execution/file-utils.ts:144]: I guess we can also use the `!isLocalFolder` directly on lines 16 and 47 and wipe out `isSingleFile` all together?

File: src/cli/commands/test/iac-local-execution/file-utils.ts [hunk 16]
```diff
@@ -101,3 +103,42 @@ export function computePaths(
...
 }
+
+/**
+ * makeFileAndDirectoryGenerator is a generator function that helps walking the directory and file structure of this pathToScan
+ * @param root
+ * @param maxDepth? - An optional `maxDepth` argument can be provided to limit how deep in the file tree the search will go.
+ * @returns {Generator<object>} - a generator which yields an object with directories or paths for the path to scan
+ */
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function* makeFileAndDirectoryGenerator(root = '.', maxDepth?: number) {
```
Comment (ipapast) [inline src/cli/commands/test/iac-local-execution/file-utils.ts:114]: reminder to fix the depth as this is wrong based on https://github.com/snyk/cli/pull/3041/files#r841972315

**Review by ipapast — COMMENTED** [review]
it looks great. 💟  I tried a few scenarios locally, but our edge cases should be covered by our acceptance tests now.

I think if we fix the detectionDepth this PR is ready to get merged so that it doesn't become stale. we can refactor again if needed as a follow up

**Review by ipapast — APPROVED** [review]
